Improvement: meta-data handling

ilya-causalens opened the pull request:
## Description and Motivation
<!--- Describe your changes, and the motivation behind them, in detail -->
- Added metadata handling system to the `cai_causal_graph.interfaces.HasMetadata` class. All extending classes should use
  this system to parse and set their metadata.
- Added `cai_causal_graph.graph_components.TimeSeriesEdge` class, which is used as an edge class by `cai_causal_graph.time_series_causal_graph.TimeSeriesCausalGraph`.
- Added metadata to `cai_causal_graph.time_series_causal_graph.TimeSeriesCausalGraph` and `cai_causal_graph.causal_graph.CausalGraph`.
- Ensured consistent metadata handling. Metadata passed at construction to `cai_causal_graph.graph_components.TimeSeriesNode`,
  `cai_causal_graph.graph_components.Node`, `cai_causal_graph.graph_components.TimeSeriesEdge`, `cai_causal_graph.graph_components.Edge`,
  `cai_causal_graph.time_series_causal_graph.TimeSeriesCausalGraph` and `cai_causal_graph.causal_graph.CausalGraph` is shallow-copied. Any
  metadata is deepcopied when constructing `cai_causal_graph.time_series_causal_graph.TimeSeriesCausalGraph` and `cai_causal_graph.causal_graph.CausalGraph`
  from dictionary.
- Ensured that `cai_causal_graph.time_series_causal_graph.TimeSeriesCausalGraph.replace_node` performs similarly to
  `cai_causal_graph.causal_graph.CausalGraph.replace_node`, meaning that if any additional information (such as metadata)
  is specified, it is used to overwrite corresponding information in the constructed node.
- Generalized `cai_causal_graph.causal_graph.CausalGraph.__eq__` to check for the class of the instance itself, enabling
  to reuse this method by extending classes.

## Additional Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here (e.g. Jira). -->
This PR is mainly aimed at two things:
1) Consistency between implementations of methods by `TimeSeriesCausalGraph` and `CausalGraph`.

Currently, there are minor inconsistencies between implementations of certain methods. For example, in some cases, metadata is copied, but not in others. `TimeSeriesCausalGraph.replace_node` does not provide a simple way to overwrite metadata, while `CausalGraph.replace_node` does. etc.

2) Consistency and extendability of the handling of metadata.

Metadata handling has become problematic when extending `CausalGraph` classes. This is because in most cases, extending classes simply require adding some metadata to the class (which can be easily accessed through a property). This should be simple, and should not require overwritting any complex logic.

The metadata system built in this PR is generally a starting point. It provides a skeleton for handling metadata, which can be extended in the future. For example, it would be nice to enable automatically adding properties to the class based on the metadata schema, or providing callable default values. These improvements can be added later.

## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Improvement (non-breaking change which improves functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Refactor (moving code around, general refactoring improvements)
- [ ] Documentation (documentation)


## PR Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->

- [ ] I have reviewed my own PR? (review the PR as you are reviewing someone else's PR)
- [ ] I have implemented all requirements? (see JIRA, project documentation)
- [ ] I am affecting someone else's work? If so: I have added those people as reviewers?
- [ ] I have added comments to all the bits that are hard to follow
- [ ] At a bare minimum, I tested this manually
- [ ] I have added unit test(s)
- [ ] Is this a bugfix? If so have I added a regression test?
- [ ] Does this PR satisfy the [one PR, one concern](https://medium.com/@fagnerbrack/one-pull-request-one-concern-e84a27dfe9f1) rule?
- [ ] If needed, documentation has been added/updated?
- [ ] I have updated the appropriate changelog with a line for my changes
- [ ] If this PR breaks reproducibility, have I added a note in the changelog explaining the break in reproducibility
      and its extent?

## Screenshots (if appropriate):

andrew-causalens commented on cai_causal_graph/time_series_causal_graph.py:
```diff
@@ -208,7 +185,7 @@ def get_stationary_graph(self) -> TimeSeriesCausalGraph:
 
         # now extend the minimal graph to the current max and min lag to match the current graph
-        self._stationary_graph = minimal_graph.extend_graph(-neg_lag, pos_lag)
+        self._stationary_graph = minimal_graph.extend_graph(-neg_lag, pos_lag, include_all_parents=False)
```
Was the plan to change the default to `True` for `include_all_parents` or did MLE/Dara land on a different solution? I saw no one replied to your question on team_dara.

ilya-causalens replied:
I am still waiting for them. Claudia said she is fine with it, but I said I'll wait for her to give me a formal reply. I figured it can be done as a separate PR (before we release a new minor).

andrew-causalens replied:
Sounds good. Just noticed this was set explicitly so it was hinting that default was changing. If we do change it then the docs actually need an update as they give some examples of how extend works.

andrew-causalens reviewed (approved):
Great stuff @ilya-causalens. I think I found one bug in `TimeSeriesEdge` but rest is just minor comments.

maxelliott-causalens commented on cai_causal_graph/graph_components.py:
```diff
@@ -666,6 +584,66 @@ def to_dict(self, include_meta: bool = True) -> dict:
+    """
+    Class defining a time-series edge.
+
+    Time series edge is equivalent to a `cai_causal_graph.graph_components.Edge` class, except that it only
```
**A** time-series edge...

Would add a hyphen to be consistent with format you have used elsewhere in the docstring

maxelliott-causalens commented on cai_causal_graph/graph_components.py:
```diff
@@ -666,6 +584,66 @@ def to_dict(self, include_meta: bool = True) -> dict:
+            if source.time_lag > destination.time_lag:
+                _destination = source
+                source = destination
+                destination = _destination
```
Completely unnecessary optimization but I discovered recently that you can do this to swap variables in one line 🙃
```
source, destination = destination, source
```

ilya-causalens commented on cai_causal_graph/metadata_handler.py:
```diff
@@ -0,0 +1,38 @@
+"""
+Copyright (c) 2024 by Impulse Innovations Ltd. Private and confidential. Part of the causaLens product.
+"""
+from dataclasses import dataclass
```
I prefer them for their simplicity. IMO if all that is needed is a `dataclass`, then just use that. If anything more advanced is needed, use `pydantic`. I can change it if you have a strong opinion?

andrew-causalens replied:
No strong opinion. I think we have mixed use throughout product so just wanted to ask. Will mark resolved.

ilya-causalens commented on changelog.md:
```diff
@@ -1,5 +1,25 @@
+> and `cai_causal_graph.causal_graph.CausalGraph`, which means their serialized state has changed.
+
+- Added metadata handling system to the `cai_causal_graph.interfaces.HasMetadata` class. All extending classes should use
+  this system to parse and set their metadata.
```
Ah, great catch. Completely forgot!

maxelliott-causalens commented on tests/test_time_series_graph.py:
```diff
@@ -2001,3 +2002,56 @@ def test_lagged_floating_nodes_correctly_added_to_minimal_graph(self):
+        ...
+
+    def test_get_stationary_deepcopies_meta(self):
+        ...
```
Need implementing?

maxelliott-causalens reviewed (approved):
Nice job! See you've dealt with CausalModelGraph as well in the PR so cheers for that!

ilya-causalens commented:
Hey @patricia-causalens - I have added you as a reviewer. Appreciate that it is a fairly large PR, but the main change is the addition of `'meta'` to `CausalGraph`/`TimeSeriesCausalGraph`, which changes its serialization. I have also added you to the [PR](https://github.com/causalens/cai-modeling/pull/248) in `cai-modeling` which extends this change to `CausalModelGraph` and `TimeSeriesCausalModelGraph`.

patricia-causalens reviewed (approved):
Thanks, LGTM
